### PR TITLE
add parsing of types to functions

### PR DESCRIPTION
**Pull request by jessfraz**
This is a start of making generating docs from kcl functions written in kcl easier.

It allows us to optionally declare types in a function.

What this does:

- will allow me to introspect the type so we can generate docs for the args

What this does not do:

- Any actual compile time or runtime validation of said types.

I figure we can tackle the latter later on should we want. But right now this is merely for documentation purposes.

Next steps (in different prs):

- declaring return type for functions

cc @lf94 @adamchalmers

### CODE REVIEW

**Comment by lf94** [comment]
1. I don't see support for multi-type arguments, ex. `fn f(a: number[] | { prop1: string, prop: number }, b: string)`
2. `number` isn't fine-grained enough. Many functions fail if they are not for example, given a float

Otherwise looks great to me

**Comment by jessfraz** [comment]
@lf94 ah yeah i just added off the top of head figured id add shit as we hit it but ill fix

**Comment by jessfraz** [comment]
we actaully parse all ints as floats so every number is a float

**Comment by jessfraz** [comment]
we dont have enums in kcl yet so i dont even know how multi type would work

**Comment by lf94** [comment]
> we actaully parse all ints as floats so every number is a float

Ah ok. In Grackle this isn't the case _yet_. Nevermind then about the finer `number` type

File: src/wasm-lib/kcl/src/ast/types.rs [hunk 1]
```diff
@@ -2753,6 +2753,41 @@ async fn execute_pipe_body(
...
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, JsonSchema, Bake)]
+#[databake(path = kcl_lib::ast::types)]
+#[serde(tag = "type")]
+pub enum FnArgType {
+    /// A primitive type.
+    Primitive(FnArgPrimitive),
+    // An array of a primitive type.
+    Array(FnArgPrimitive),
```
Comment (lf94) [inline src/wasm-lib/kcl/src/ast/types.rs:2784]: Will this allow T[][][]... ? (KCL supports nested arrays)

Comment (jessfraz) [reply src/wasm-lib/kcl/src/ast/types.rs:2784]: sure i can do that i dont think we do that in any std lib fn tho

Comment (jessfraz) [reply src/wasm-lib/kcl/src/ast/types.rs:2784]: on second thought that gets super recursive and until we need it id rather punt, im not that advanced at winnow yet

**Pull request merged**